### PR TITLE
Save custom tensor API

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- Evaluate PR only after #267 
- New API `save_custom_tensor`
- Tests to support the new API

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/keras.py [hunk 1]
```diff
@@ -319,6 +329,8 @@ def _prepare_non_layer_tensors(self):
         for coll in [
             self.get_collection(name=CollectionKeys.OPTIMIZER_VARIABLES),
             self.get_collection(name=CollectionKeys.GRADIENTS),
+            self.get_collection(name=CollectionKeys.OUTPUTS),
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:332]: can you provide details in description aorund these changes?

File: smdebug/tensorflow/keras.py [hunk 2]
```diff
@@ -584,14 +641,13 @@ def _on_any_batch_end(self, batch, mode, logs=None):
         if not is_tf_version_2x() or (is_tf_version_2x() and not tf.executing_eagerly()):
             self._remove_fetches_and_callbacks(mode)
 
+        self._save_tensors_post_step(batch, logs)
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:644]: any reason for moving from below to here?

File: smdebug/tensorflow/keras.py [hunk 3]
```diff
@@ -366,6 +374,56 @@ def _add_metric(self, metric_name, metric_value: tf.Tensor = None):
...
+                tensor_ref = TensorRef.from_non_graph_var(tensor_name)
+                tensor_refs.append((tensor_ref, t))
+        else:
+            tensor_ref = TensorRef.from_non_graph_var(tensor_name)
+            tensor_refs.append((tensor_ref, tensor_value))
+
+        for tensor_ref, t in tensor_refs:
+            for collection in collections_to_write:
+                collection.set_tensor_ref(tensor_ref)
+            self._save_for_tensor(tensor_name, t, check_before_write=False)
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:399]: how will save_custom_tensor work for non keras , pytorch,mxnet,xgboost?

File: smdebug/tensorflow/keras.py [hunk 4]
```diff
@@ -95,6 +100,11 @@ def _is_not_supported(self):
                 self._hook_supported = False
         return not self._hook_supported
 
+    def should_save_global_step_for_mode(self, mode: str):
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py:103]: This is used in AWF TF to decide if a certain step should be saved.

**Pull request closed without merging**